ESQL: Support ST_CONTAINS and ST_WITHIN

craigtaverner opened the pull request:
This was a relatively small update on top of the main work on ST_INTERSECTS at https://github.com/elastic/elasticsearch/pull/104907. Most of the code is either test code, or generated code, since the foundations in `SpatialRelatesFunction` and related classes covered most of the needs.

We do both `ST_CONTAINS` and `ST_WITHIN` at the same time, due to the commutative relationship between these two. In particular we want to move literals to the right field, to both minimise the number of evaluators and simplify Lucene push-down. This is only possible if we can also swap `ST_CONTAINS` with `ST_WITHIN`, since:

```
  ST_CONTAINS(a, b) = ST_WITHIN(b, a)
```

Tasks:

* [x] Implement SpatialContains, and add it to PlanNamedTypes and EsqlFunctionRegistry
* [x] Implement SpatialContainsTests and make all tests pass:
  * [x] Fix multi-components (See [LatLongShape.makeContainsGeometryQuery](https://github.com/apache/lucene/blob/c78533c53cba8072cecabf62269ec8f0bfb35bed/lucene/core/src/java/org/apache/lucene/document/LatLonShape.java#L355) and [ShapeDocValuesQuery.getContainsWeight](https://github.com/elastic/elasticsearch/blob/cf5fbfc857b36438baf357a31a7c2bc960111797/server/src/main/java/org/elasticsearch/lucene/spatial/ShapeDocValuesQuery.java#L165)
  * [x] Fix rectangles crossing dateline (see Lucene's Rectangle2D)
* [x] Implement `ST_WITHIN` to avoid needing too many evaluators (support inverting between contains and within)
  * [x] Implement SpatialWithin, and add it to PlanNamedTypes and EsqlFunctionRegistry
  * [x] Implement LogicalPlanOptimizer rule to invert between contains and within when left argument is constant and right is not
  * [x] Implement SpatialWithinTests and make all tests pass:
* [x] Implement integration tests (using existing airports and airport_city_boundaries datasets)
  * [x] Basic tests loading data from airports and city_boundaries and performing predicates
  * [x] Tests together with ST_CENTROID for doc-values support
  * [x] tests with EVAL with and without centroid
* [x] Develop Lucene-pushdown in `LocalPhysicalPlanOptimizer` and `PhysicalPlanOptimizerTests`
  * [x] Validate interaction with point doc-values (`ST_CONTAINS` and `ST_WITHIN` before `ST_CENTROID`)
* [x] Write documentation
* [x] Fix contains/within for Cartesian (currently disabled)
  * [x] Add more comprehensive multi-polygon integration tests for cartesian intersects/contains/within
  * [x] Re-enable lucene push-down for contains/within with cartesian data
  * [x] Fix any remaining issues discovered from these integration tests in lucene push-down
* [ ] Either implement `ST_DISJOINT` or create an issue to do that

craigtaverner committed:
WIP Started adding ST_CONTAINS

elasticsearchmachine commented:
Hi @craigtaverner, I've created a changelog YAML for you.

craigtaverner committed:
Add generated evaluators

craigtaverner committed:
Reduced warnings and use correct evaluators

craigtaverner committed:
Refactored tests to remove duplicate code, and fixed Contains/multi-components

craigtaverner committed:
Gradle build disallows using getDeclaredField

craigtaverner committed:
Fixed cases where rectangles cross the dateline

craigtaverner committed:
Fixed meta function tests

craigtaverner committed:
Added ST_WITHIN to support inverting ST_CONTAINS

If the ST_CONTAINS is called with the constant on the left, we either have to create a lot more Evaluators to cover that case, or we have to invert it to ST_WITHIN. This inversion was a much easier option.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -175,6 +177,45 @@ protected static List<Batch<LogicalPlan>> rules() {
+                return srf;
+            }
+            if (srf.left().foldable() && srf.right().foldable() == false) {
+                // If we have a constant on the left, and a field on the right, invert the function
```
1. Why is this necessary/helpful?
2. Is this really equivalent in all cases? It looks to me like inverting e.g. `SpatialWithin` or `SpatialContains` produces a non-equivalent condition. Don't we need to negate the whole thing for this to work?

alex-spies replied:
Scratch No.2 - I just realized that within/contains gets replaced by the other, resp.

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialRelatesFunction.java:
```diff
@@ -203,13 +203,17 @@ public boolean rightDocValues() {
+
+    public abstract boolean isCommutative();
+
+    public abstract SpatialRelatesFunction invert();
```
I don't think we need this - `BinaryScalarFunction` already has `replaceChildren` which should do what you need.

alex-spies replied:
Woops, never mind - `replaceChildren` won't let you exchange the left and right doc values, sorry!

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -175,6 +177,45 @@ protected static List<Batch<LogicalPlan>> rules() {
+            }
+            if (srf.left().foldable() && srf.right().foldable() == false) {
+                // If we have a constant on the left, and a field on the right, invert the function
+                return srf.invert();
```
This looks very close to `LiteralsOnTheRight`. Is there a chance we can just use that one? E.g. by having `SpatialRelatesFunction` inherit from `BinaryOperator` rather than from `BinaryScalarFunction` (that may be a big ask, though, not sure that would work); or by generalizing `LiteralsOnTheRight` to handle both cases?

craigtaverner replied:
I did investigate BinaryOperator for the original SpatialIntersects function, but it turns out BinaryOperator has a lot of assumptions that are not applicable here, so I backed out of that approach. I'll take a peek at LiteralsOnTheRight, to see if it is amenable to this.

craigtaverner committed:
Simplify inversion logic

alex-spies commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -177,42 +177,15 @@ protected static List<Batch<LogicalPlan>> rules() {
-                return srf.invert();
-            }
-            return srf;
+        protected SpatialRelatesFunction rule(SpatialRelatesFunction function) {
+            return function.surrogate();
```
Looks good in the new version!

craigtaverner replied:
Thanks, and I am adding a comment explaining the similarity to the two other approaches, and the decision to go with this approach for now.

craigtaverner committed:
Add comment on choice of surrogate approach

craigtaverner committed:
Add unit tests and missing fold() function

craigtaverner committed:
Simple code cleanup

craigtaverner committed:
Add integration tests for literals

craigtaverner committed:
Add more integration tests based on actual data

craigtaverner committed:
Generated documentation files

craigtaverner committed:
Add documentation

craigtaverner committed:
Fixed failing function count test

craigtaverner committed:
Add tests that push-to-source works for ST_CONTAINS and ST_WITHIN

craigtaverner committed:
Test more combinations of WITH/CONTAINS and literal on right and left

This also verifies that the re-writing of CONTAINS to WITHIN or vice versa occurs when the literal is on the left.

craigtaverner committed:
test that physical planning also handles doc-values from STATS

craigtaverner committed:
Added more tests for WITHIN/CONTAINS together with CENTROID

This should test the doc-values for points.

craigtaverner committed:
Add cartesian_point tests

craigtaverner committed:
Add cartesian_shape tests

nik9000 commented on docs/reference/esql/functions/st_contains.asciidoc:
```diff
@@ -0,0 +1,34 @@
+[%header.monospaced.styled,format=dsv,separator=|]
+|===
+include::{esql-specs}/spatial.csv-spec[tag=st_contains-airport_city_boundaries-results]
+|===
```
You can *generally* just link to `esql/functions/layout/st_contains.asciidoc` - especially if you add an `@Example` to `@FunctionInfo` annotation. Then it *should*  spit out pretty much the same thing as this. You may want a `note` for the "The second parameter must al..." part.

craigtaverner replied:
Cool! I'm trying the @Example option, but will leave the rest of the options alone in this PR and play with this more in later work.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialContains.java:
```diff
@@ -0,0 +1,280 @@
+    public SpatialContains(
+        Source source,
+        @Param(
+            name = "geomA",
```
I'd use a full word. This will be used by Kibana for autocomplete, but I am not sure if as is or they will give it a proper name.

craigtaverner replied:
I used `geomA` and `geomB` because that is exactly what PostGIS used, and I kind-of liked the idea that our docs look a little similar to PostGIS docs. But Kibana obviously takes precedence. I could ask them about this.

astefan commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/SpatialWithin.java:
```diff
@@ -0,0 +1,245 @@
+            description = "Geometry column name or variable of geometry type"
+        ) Expression left,
+        @Param(
+            name = "geomB",
```
Same here about the parameter names.

astefan reviewed (approved):
Two minor comments. LGTM

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/spatial/LuceneComponent2DUtils.java:
```diff
@@ -0,0 +1,91 @@
+    /**
+     * This method is based on LatLonGeometry.create, but returns an array of Component2D objects for multi-component geometries.
+     */
+    public static Component2D[] createLatLonComponents(LatLonGeometry... latLonGeometries) {
```
Move towards InvalidArgumentException since those have proper HTTP codes.

craigtaverner replied:
Right now, all the spatial functions use IllegalArgumentException, and that was based on copying from other functions. I think I should move the whole lot over to InvalidArgumentException in one go, in a separate PR, if that is OK for you?

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanNamedTypes.java:
```diff
@@ -1485,6 +1489,28 @@ static void writeIntersects(PlanStreamOutput out, SpatialIntersects intersects)
+    }
+
+    static void writeContains(PlanStreamOutput out, SpatialContains intersects) throws IOException {
+        List<Expression> fields = intersects.children();
+        assert fields.size() == 2;
+        out.writeExpression(fields.get(0));
+        out.writeExpression(fields.get(1));
```
```
out.writeExpression(intersects.left())
out.writeExpression(intersects.right())
```
No need to do assertion or generic extractions.

craigtaverner replied:
Done, and merged the three write functions into one, since they are identical. But we still need three read functions to instantiate the correct classes.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/io/stream/PlanNamedTypes.java:
```diff
@@ -1485,6 +1489,28 @@ static void writeIntersects(PlanStreamOutput out, SpatialIntersects intersects)
+        return new SpatialWithin(Source.EMPTY, in.readExpression(), in.readExpression());
+    }
+
+    static void writeWithin(PlanStreamOutput out, SpatialWithin intersects) throws IOException {
+        List<Expression> fields = intersects.children();
+        assert fields.size() == 2;
+        out.writeExpression(fields.get(0));
+        out.writeExpression(fields.get(1));
+    }
```
Same as above.

costin commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java:
```diff
@@ -174,6 +176,25 @@ protected static List<Batch<LogicalPlan>> rules() {
+     * this in either direction, reducing the number of rules, but for now,
+     * it's a separate rule to reduce the risk of unintended interactions with other rules.
+     */
+    static class SubstituteSpatialSurrogates extends OptimizerRules.OptimizerExpressionRule<SpatialRelatesFunction> {
```
Move it under SubstituteSurrogate and please raise an issue to indicate that this needs to either be updated based on either LiteralsOnTheRight OR be incorporated into SubstituteSurrogates

craigtaverner replied:
Moved and created issue at https://github.com/elastic/elasticsearch/issues/106763

costin reviewed (approved):
LGTM

craigtaverner committed:
Disable Lucene-push-down for CARTESIAN data

This is a limitation in Lucene, which we could address as a performance optimization in a future PR, but since it probably requires Lucene changes, it cannot be done in this work.

craigtaverner committed:
Fix doc links

craigtaverner committed:
Added test data and tests for cartesian multi-polygons

Testing INTERSECTS, CONTAINS and WITHIN with multi-polydon fields

craigtaverner committed:
Use required features for spatial points, shapes and centroid

craigtaverner committed:
8.13.0 is not yet historical version

This needs to be reverted as soon as 8.13.0 is released

craigtaverner committed:
Added st_intersects and st_contains_within 'features'

craigtaverner committed:
Code review updates

craigtaverner committed:
Re-enable lucene push-down

craigtaverner committed:
Added more required_features

craigtaverner committed:
Fix point contains non-point

craigtaverner committed:
Fix point contains point

craigtaverner committed:
Re-enable lucene push-down in tests too

Forgot to change the physical planner unit tests after re-enabling lucene push-down

craigtaverner committed:
Generate automatic docs

craigtaverner committed:
Use generated examples docs

craigtaverner committed:
Generated examples use '-result' prefix (singular)

iverase commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/querydsl/query/SpatialRelatesQuery.java:
```diff
@@ -231,9 +231,9 @@ private static org.apache.lucene.search.Query pointShapeQuery(
+                return new MatchNoDocsQuery();
             }
             final Consumer<ShapeType> checker = t -> {
                 if (t == ShapeType.POINT || t == ShapeType.MULTIPOINT || t == ShapeType.LINESTRING || t == ShapeType.MULTILINESTRING) {
```
I wonder if this checker is correct here. I think in cartesian we can generate any geometry because there is no loosy quantization so everything is in theory supported.

craigtaverner committed:
Mark spatial functions as preview/experimental
